[RF] Make it possible to store global observables in `RooAbsData` and consider these global observables in `RooAbsPdf::fitTo`

guitargeek opened the pull request:
This PR implements the possibility to store global observables in RooFit datasets, and makes the necessary changes in the `RooAbsPdf::fitTo` code path to consider the global observables in the data if available. If one wants to restore the old behavior of taking the global observable values from the model even if they are stored in the data, one can use the new `GlobalObservablesSource` command argument. 

Unit tests for all new functionality is also implemented.

Please find the more information in the commit messages.

Ideas for future developments in future PRs:
1. make it possible to specify the global observables when generating a dataset:
    `model.generate(x, 1000, GlobalObservables(g))`
2. make it possible to also sample global observable values when generating a toy dataset:
    `model.generate({x, g}, 1000, GlobalObservables(g))`
3. Add a tutorial to show all the new functionality related to global observables

Closes https://github.com/root-project/root/issues/8123.

hageboeck reviewed (commented):
Looks very good already, but I added quite some comments.

Maybe think about separating code changes and adding tests in the same commit. If you e.g. want to try whether a specific change breaks a test, you could just revert the code part without removing the test.
Not important for this PR, but could be useful as a general strategy.

hageboeck commented on roofit/roofitcore/inc/RooAbsData.h:
```diff
@@ -297,6 +297,9 @@ class RooAbsData : public TNamed, public RooPrintable {
 
   static StorageType getDefaultStorageType();
 
+  RooArgSet const* getGlobals() const { return _globalObservables.get(); }
```
While reading through the rest of the commits, I actually noticed that the function name isn't so great. It looks like dealing with c++ globals. It should at least mention `GlobalObservables`.

guitargeek replied:
You're right, I'll call it `getGlobalObservables` then.

guitargeek commented on roofit/roofitcore/inc/RooAbsPdf.h:
```diff
@@ -159,6 +159,32 @@ class RooAbsPdf : public RooAbsReal {
                               const RooCmdArg& arg6=RooCmdArg::none(),  const RooCmdArg& arg7=RooCmdArg::none(), const RooCmdArg& arg8=RooCmdArg::none()) ;
   virtual RooFitResult* fitTo(RooAbsData& data, const RooLinkedList& cmdList) ;
 
+  /// Configuration struct for RooAbsPdf::minimizeNLL with all the default
```
Ok, thanks!

guitargeek commented on roofit/roofitcore/inc/RooAbsTestStatistic.h:
```diff
@@ -51,6 +51,7 @@ class RooAbsTestStatistic : public RooAbsReal {
     bool cloneInputData = true;
     double integrateOverBinsPrecision = -1.;
     bool binnedL = false;
+    bool takeGlobsFromData = false;
```
Thanks!

guitargeek commented on roofit/roofitcore/src/RooAbsPdf.cxx:
```diff
@@ -1420,6 +1420,101 @@ RooFitResult* RooAbsPdf::fitTo(RooAbsData& data, const RooCmdArg& arg1, const Ro
+///            the RooMinimizer. These are a subset of the options that you can
+///            also pass to RooAbsPdf::fitTo via the RooFit command arguments.
+std::unique_ptr<RooFitResult> RooAbsPdf::minimizeNLL(RooAbsReal & nll,
+        RooAbsData& data, MinimizerConfig const& cfg) {
```
Yes, you are right is still compiles with `const` :+1:

guitargeek commented on README/ReleaseNotes/v626/index.md:
````diff
@@ -96,6 +96,48 @@ RooFit now contains two RDataFrame action helpers, `RooDataSetHelper` and `RooDa
+Usage example for a model with global observables `g1` and `g2`:
+```C++
+auto data = model.generate(x, 1000); // data has only the single observables x
+data->setGlobals(g1, g2); // now, data also stores a snapshot of g1 and g2
````
You're right, as mentioned in another comment, I'll rename this to `getGlobalObservables`.

guitargeek commented on roofit/roofitcore/inc/RooGlobalFunc.h:
```diff
@@ -239,10 +239,14 @@ RooCmdArg SplitRange(Bool_t flag=kTRUE) ;
+  return RooCmdArg("GlobalObservables",0,0,0,0,0,0,0,0,0,0,
+          &RooCmdArg::take(RooArgSet{std::forward<Args_t>(argsOrArgSet)...}));
+}
+RooCmdArg GlobalObservablesSource(const char* sourceName);
```
Exactly, enums are more efficient but are not very pythonic. I went for strings, because then in Python it will work like `model.fitTo(data, GlobalObservablesSource="model")`.

guitargeek commented on roofit/roofitcore/test/testGlobalObservables.cxx:
```diff
@@ -0,0 +1,277 @@
+                                       RooCmdArg const &arg4 = RooCmdArg::none())
+   {
+      using namespace RooFit;
+      return std::unique_ptr<RooFitResult>(
+         model.fitTo(data, Save(), Verbose(false), PrintLevel(-1), arg1, arg2, arg3, arg4));
```
Yes I thought so too, but is doesn't:
```
error: could not convert ‘{(& model)->RooAbsPdf::fitTo((* & data), RooFit::Save(Bool_t)(), RooFit::Verbose(Bool_t)(), RooFit::PrintLevel(Int_t)(), (* & arg1), (* & arg2), (* & arg3), (* & arg4), (* & RooCmdArg::none()))}’ from ‘<brace-enclosed initializer list>’ to ‘std::unique_ptr<RooFitResult>’
  125 |       return {model.fitTo(data, Save(), Verbose(false), PrintLevel(-1), arg1, arg2, arg3, arg4)};
      |                                                                                                ^
      |                                                                                                |
      |                                                                                                <brace-enclosed initializer list>
```

guitargeek commented on README/ReleaseNotes/v626/index.md:
````diff
@@ -96,6 +96,48 @@ RooFit now contains two RDataFrame action helpers, `RooDataSetHelper` and `RooDa
+With this release, a mechanism is added to store a snapshot of global observables in any `RooDataSet` or `RooDataHist`.
+For toy studies where the global observables assume a different values for each toy, the bookkeeping of the set of global observables and in particular their values is much easier with this change.
+
+Usage example for a model with global observables `g1` and `g2`:
+```C++
+auto data = model.generate(x, 1000); // data has only the single observables x
+data->setGlobals(g1, g2); // now, data also stores a snapshot of g1 and g2
+
+// If you fit the model to the data, the global observables and their values
+// are taken from the dataset:
+model.fitTo(*data);
+
+// You can still define the set of global observables yourself, but the values
+// will be takes from the dataset if available:
+model.fitTo(*data, GlobalObservables(g1, g2));
+ 
+// To force `fitTo` to take the global observable values from the model even
+// though they are in the dataset, you can use the new `GlobalObservablesSource`
+// command argument:
+model.fitTo(*data, GlobalObservables(g1, g2), GlobalObservablesSource("model"));
+// The only other allowed value for `GlobalObservablesSource` is "data", which
+// corresponds to the new default behavior explained above.
+```
````
Yes, I would also like to make a tutorial about these global observables. but maybe not in this PR but only when `RooAbsPdf::generate` can also deal with global observables. I have added it to the checklist of things to do next in the PR description.

guitargeek commented on roofit/roofitcore/test/testGlobalObservables.cxx:
```diff
@@ -0,0 +1,277 @@
+
+   // verify that taking the global observable values from data has not changed
+   // the values in the model
+   EXPECT_EQ(gmVaryVal, gm.getVal());
+   EXPECT_EQ(gsVaryVal, gs.getVal());
```
Okay! I have added messages to all `EXPECT`.

guitargeek commented on roofit/roofitcore/src/RooAbsPdf.cxx:
```diff
@@ -1420,6 +1420,101 @@ RooFitResult* RooAbsPdf::fitTo(RooAbsData& data, const RooCmdArg& arg1, const Ro
+        RooAbsData& data, MinimizerConfig const& cfg) {
+
+  // Determine if the dataset has weights
+  bool weightedData = data.isNonPoissonWeighted() ;
```
Right, it was copied code. I will remove the space.

guitargeek commented on roofit/roofitcore/inc/RooConstraintSum.h:
```diff
@@ -28,7 +28,7 @@ class RooConstraintSum : public RooAbsReal {
 
   RooConstraintSum() {}
-  RooConstraintSum(const char *name, const char *title, const RooArgSet& constraintSet, const RooArgSet& paramSet) ;
+  RooConstraintSum(const char *name, const char *title, const RooArgSet& constraintSet, const RooArgSet& paramSet, bool takeGlobsFromData=false) ;
```
Okay good point, I'll go for the longer but more expressive name `takeGlobalObservablesFromData`.

guitargeek commented on roofit/roofitcore/src/RooConstraintSum.cxx:
```diff
@@ -91,6 +91,16 @@ Double_t RooConstraintSum::evaluate() const
 
 
+////////////////////////////////////////////////////////////////////////////////
+/// Redirect the global observables to the ones from the dataset if applicable.
```
Ok, I have written a description with some more detail:
```
/// Replace the varaiables in this RooConstraintSum with the global observables
/// in the dataset if they match by name. This function will do nothing if this
/// RooConstraintSum is configured to not use the global observables stored in
/// datasets.
```

guitargeek commented on roofit/roofitcore/src/RooGlobalFunc.cxx:
```diff
@@ -222,10 +222,8 @@ namespace RooFit {
   RooCmdArg Constrain(RooArgSet && params) { return Constrain(RooCmdArg::take(std::move(params))); }
-  RooCmdArg GlobalObservables(const RooArgSet& globs)    { return RooCmdArg("GlobalObservables",0,0,0,0,0,0,0,0,0,0,&globs) ; }
-  RooCmdArg GlobalObservables(RooArgSet && globs) { return GlobalObservables(RooCmdArg::take(std::move(globs))); }
+  RooCmdArg GlobalObservablesSource(const char* sourceName) { return {"GlobalObservablesSource",0,0,0,0,sourceName,0,0,0}; }
```
I'm not sure if these functions should really be documented here. A given command argument can be used in different places with different effects, so the documentation should better go to the docs of the function that uses the command argument. For example, the `GlobalObservables` command acts different in `RooAbsPdf::fitTo` or `RooDataSet::RooDataSet`.

I agree that the situation with the documentation of command arguments needs to be improved, but maybe not in this PR.

guitargeek commented on README/ReleaseNotes/v626/index.md:
```diff
@@ -96,6 +96,48 @@ RooFit now contains two RDataFrame action helpers, `RooDataSetHelper` and `RooDa
 
+### Storing global observables in RooFit datasets
+
+RooFit groups model variables into *observables* and *parameters*, depending on if their values are stored in the dataset.
+For fits with parameter constraints, there is a third kind of variables, called *global observables*.
+These represent the results of auxiliary measurements that constrain the nuisance parameters.
```
Good point :+1:

hageboeck commented on roofit/roofitcore/src/RooConstraintSum.cxx:
```diff
@@ -88,6 +93,19 @@ Double_t RooConstraintSum::evaluate() const
 
 
+////////////////////////////////////////////////////////////////////////////////
+/// Replace the varaiables in this RooConstraintSum with the global observables
```
```suggestion
/// Replace the variables in this RooConstraintSum with the global observables
```

guitargeek replied:
:+1: :

hageboeck commented on roofit/roofitcore/test/CMakeLists.txt:
```diff
@@ -44,3 +44,4 @@ ROOT_ADD_GTEST(testRooProductPdf testRooProductPdf.cxx LIBRARIES RooFitCore)
 ROOT_ADD_GTEST(testNaNPacker testNaNPacker.cxx LIBRARIES RooFitCore)
 ROOT_ADD_GTEST(testRooSimultaneous testRooSimultaneous.cxx LIBRARIES RooFitCore RooFit)
 ROOT_ADD_GTEST(testRooGradMinimizerFcn testRooGradMinimizerFcn.cxx LIBRARIES RooFitCore)
+ROOT_ADD_GTEST(testGlobalObservables testGlobalObservables.cxx LIBRARIES RooFitCore RooFit)
```
```suggestion
ROOT_ADD_GTEST(testGlobalObservables testGlobalObservables.cxx LIBRARIES RooFit)
```
I thought for a long time that you have to mention both, but CMake's transitive dependencies should allow for only mentioning `RooFit`. If it doesn't work, the CMake config needs to be fixed.

hageboeck replied:
Oh, but on second look:
This introduces a dependency to RooFit, i.e., the entire set of libRooFit.so files needs to be compiled before you can run this test.

guitargeek replied:
True, I tried to avoid this now by using `RooGenericPdf` instead of `RooPoisson` and `RooGaussian`. However, the test gets slowed down by the numeric integration: from 185 ms to about 1.5 seconds. So while I agree that it would be nice to get rid of this dependency, I don't think we should do something about it in this PR. There is also `testRooAbsPdf` and `testRooSimultaneous` that still depend on RooFit too, and I'll try to solve this dependency problem for all of the tests at a later point.

hageboeck reviewed (commented):
I found two more things. 🙂

guitargeek commented:
Hi @hageboeck, thanks for the review! I have addressed most of your comments and tried to justify whenever I didn't do something, which was mostly because I think it should be done in separate PRs (new tutorial, removing dependecy of RooFitCore tests on RooFit).

I think this PR implements now all the functionality requested by @will-cern in https://github.com/root-project/root/issues/8123, so it would be good if you Will could confirm this.

In a later PR, I'm then going to implement what was not required to solve the issue but what is still useful for ROOT nonetheless:
1. make it possible to specify the global observables when generating a dataset:
    `model.generate(x, 1000, GlobalObservables(g))`
2. make it possible to also sample global observable values when generating a toy dataset:
    `model.generate({x, g}, 1000, GlobalObservables(g))`
3. Add a tutorial to show all the new functionality related to global observables

lmoneta reviewed (approved):
Looks good to me.
The only comment I have is that it would be nice to show the new functionality in a RooFit tutorial where you show the old way of using global observables and the one having it in the data set.

guitargeek commented:
Thanks @lmoneta! In https://github.com/root-project/root/pull/8878#issuecomment-917953348 I have mentioned the developments that will be done related to these global observables after this PR, and creating a tutorial is part of them.
